File: backend/calc.py
"""Order total math. Keep in sync with frontend/src/calc.ts - same formulas,
two languages (backend needs totals for statistics, frontend for the invoice UI)."""
from typing import Any, Dict


def line_net(item: Dict[str, Any]) -> float:
    price = item.get("price_no_vat") or 0
    qty = item.get("ordered_qty") or 0
    discount = item.get("discount") or 0
    additional_discount = item.get("additional_discount") or 0
    total_discount = max(0, min(100, discount + additional_discount))
    return price * qty * (1 - total_discount / 100)


def line_vat(item: Dict[str, Any]) -> float:
    vat_rate = item.get("vat_rate") or 0
    return line_net(item) * (vat_rate / 100)


def compute_order_totals(order: Dict[str, Any]) -> Dict[str, float]:
    subtotal = 0.0
    vat = 0.0
    for item in order.get("items", []):
        subtotal += line_net(item)
        vat += line_vat(item)
    return {"subtotal": subtotal, "vat": vat, "grand": subtotal + vat}

File: frontend/src/calc.ts
import { Order, OrderItem } from "@/src/api";

export function effectiveDiscountPct(it: OrderItem): number {
  const supplier = it.discount ?? 0;
  const additional = it.additional_discount ?? 0;
  const factor = (1 - supplier / 100) * (1 - additional / 100);
  return (1 - factor) * 100;
}

export function lineNet(it: OrderItem): number {
  const price = it.price_no_vat ?? 0;
  const qty = it.ordered_qty ?? 0;
  const supplier = it.discount ?? 0;
  const additional = it.additional_discount ?? 0;
  return price * qty * (1 - supplier / 100) * (1 - additional / 100);
}

export function lineVat(it: OrderItem): number {
  return lineNet(it) * ((it.vat_rate ?? 0) / 100);
}

export type Totals = { subtotal: number; vat: number; grand: number };

export function computeTotals(order: Order): Totals {
  let subtotal = 0;
  let vat = 0;
  for (const it of order.items) {
    subtotal += lineNet(it);
    vat += lineVat(it);
  }
  return { subtotal, vat, grand: subtotal + vat };
}

export function money(n: number): string {
  return n.toLocaleString("sr-RS", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  });
}
